README.md
# Refazynist is a Refreshable, Lazy and Animated List

It is exactly what you are looking for. Yes, an all in one Flutter list. 

### Refreshable

It can refresh... Data can came server or local system, or what you want... 

<img src="https://themt.co/files/flutter-refazynist/refazynist-refreshable.gif"/>

### Cacheable

It can cache... Live status is saving to local system with Shared Preferences and continue from last point.

<img src="https://themt.co/files/flutter-refazynist/refazynist-cachable.gif"/>

### Lazy Loading

It can loading lazy... If you want lazy loading, you can server data or whatever

| <img src="https://themt.co/files/flutter-refazynist/refazynist-lazy.gif"/> | <img src="https://themt.co/files/flutter-refazynist/refazynist-lazy-2.gif"/> |
| :----------------------------------------------------------: | :----------------------------------------------------------: |



### Animated List

It is powerful Flutter Animated List... It can add item. It can remove item. It can clear all items.

| <img src="https://themt.co/files/flutter-refazynist/refazynist-list.gif"/> | <img src="https://themt.co/files/flutter-refazynist/refazynist-list-2.gif"/> |
| :----------------------------------------------------------: | :----------------------------------------------------------: |



### Dismissible

It can dismiss... Swipe to remove. It's simple.

<img src="https://themt.co/files/flutter-refazynist/refazynist-dismissable.gif"/>

## Usage

To use this plugin, add `refazynist` as a [dependency in your pubspec.yaml file](https://flutter.dev/docs/development/platform-integration/platform-channels).

### Example 

```dart
import 'package:flutter/material.dart';
import 'package:refazynist/refazynist.dart';
import 'package:shared_preferences/shared_preferences.dart';


void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Refazynist Dismissible Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: RefazynistDismissibleDemo(title: 'Refazynist Dismissible Demo'),
    );
  }
}

class RefazynistDismissibleDemo extends StatefulWidget {
  RefazynistDismissibleDemo({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  _RefazynistDismissibleDemoState createState() => _RefazynistDismissibleDemoState();
}

class _RefazynistDismissibleDemoState extends State<RefazynistDismissibleDemo> {

  GlobalKey<RefazynistState> refazynistKey = GlobalKey();

  int lazyCount = 5; // It's for lazy loading limit
  String sharedPreferencesName = 'dismissible_demo'; // It's for cache. Storing in Shared Preferences

  @override
  Widget build(BuildContext bContext) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Refazynist(
          key: refazynistKey,
          sharedPreferencesName: sharedPreferencesName,

          onInit: () async {
            return <dynamic>['Init item 1', 'Init item 2'];
          },

          emptyBuilder: (ewContext) {
            return Stack(
              children: <Widget>[
                ListView(),
                Center(
                  child: Wrap(
                    children: [
                      Column(
                        children: [
                          Icon(
                            Icons.warning_amber_rounded,
                            size: 60,
                            color: Colors.black26,
                          ),
                          Text ('Empty'),

                          Padding(padding: EdgeInsets.only(top: 20)),

                          ElevatedButton(
                            child: Text ('Create New'),

                            onPressed: () {
                              refazynistKey.currentState!.insertItem(0, 'Created item');
                            },
                          )
                        ],
                      )
                    ],
                  ),
                ),
              ],
            );
          },

          //
          // Refazynist: It's for refresh

          onRefresh: () async {
            lazyCount = 5; // reset lazy loading...

            await Future.delayed(Duration(seconds: 2)); // Fake internet delay

            return <dynamic>['Refresh item 0', 'Refresh item 1', 'Refresh item 2'];
          },

          //
          // Refazynist: It's for lazy load

          onLazy: () async {
            List<dynamic> lazyList = <dynamic>[];

            if (lazyCount > 0) {
              lazyCount--;
              lazyList.add ('Lazy item ' + (refazynistKey.currentState!.length() + 0).toString());
              lazyList.add ('Lazy item ' + (refazynistKey.currentState!.length() + 1).toString());
            }

            await Future.delayed(Duration(seconds: 1)); // Fake internet delay

            return lazyList;
          },

          //
          // Refazynist: itemBuilder

          itemBuilder: (item, ibContext, index, animation, type) {
            return Dismissible(
                key: Key(item),
                background: Container(
                  color: Theme.of(context).primaryColor,
                  child: Icon(Icons.delete_outline, color: Colors.white,),
                ),
                onDismissed: (_) {
                  refazynistKey.currentState!.removeItem(index, disappear: true);
                  setState(() {

                  });
                },
                child: FadeTransition(
                  opacity: animation,
                  child: Padding(
                    padding: EdgeInsets.only(left: 20, right: 20, top: 30, bottom: 30),
                    child: Row(
                      children: [
                        Expanded(child: Text ('$item')),
                        ElevatedButton(
                            onPressed: () {
                              refazynistKey.currentState!.removeItem(index);
                            },
                            child: Icon(Icons.delete_outline)
                        )
                      ],
                    ),
                  ),
                )
            );

          },

          //
          // Refazynist: removed ItemBuilder (need for Flutter's Animated List)

          removedItemBuilder: (item, ibContext, index, animation, type) {
            return FadeTransition(
              opacity: animation,
              child: Padding(
                padding: EdgeInsets.only(left: 20, right: 20, top: 30, bottom: 30),
                child: Row(
                  children: [
                    Expanded(child: Text ('$item')),
                    ElevatedButton(
                        onPressed: () {
                          refazynistKey.currentState!.removeItem(index);
                        },
                        child: Icon(Icons.delete_outline)
                    )
                  ],
                ),
              ),
            );
          }
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          showModalBottomSheet(
            context: context,
            builder: (sbsContext) {
              return ListView(
                children: [
                  ListTile(
                    title: Text('Add item to top'),
                    onTap: () {
                      refazynistKey.currentState!.insertItem(0, 'Added item ${refazynistKey.currentState!.length()}');
                    },
                  ),

                  ListTile(
                    title: Text('Remove item from top'),
                    onTap: () {
                      refazynistKey.currentState!.removeItem(0);
                    },
                  ),

                  ListTile(
                    title: Text('Clear The List'),
                    onTap: () {
                      refazynistKey.currentState!.clear();
                    },
                  ),

                  ListTile(
                    title: Text('Show Shared Preferences'),
                    onTap: () async {
                      SharedPreferences _prefs = await SharedPreferences.getInstance();

                      String? spString = _prefs.getString(sharedPreferencesName);

                      print ('Shared Preferences: $spString');

                      showModalBottomSheet(
                        context: sbsContext,
                        builder: (sbsContext2) {
                          return TextField(
                            maxLines: 15,
                            decoration: InputDecoration.collapsed(hintText: spString),
                          );
                        },
                      );
                    },
                  ),

                  ListTile(
                    title: Text('Clear Shared Preferences'),
                    onTap: () async {
                      SharedPreferences _prefs = await SharedPreferences.getInstance();

                      _prefs.setString(sharedPreferencesName, '');
                    },
                  )
                ],
              );
            },
          );
        },
        child: Icon(Icons.add),
      ),
    );
  }
}
```



CHANGELOG.md
## [0.0.4]

*  Same bug fixes

## [0.0.3]

*  New addons
*  Same bug fixes

## [0.0.2]

*  Same bug fixes

## [0.0.1]

*  Begin of everything :]

lib/src/refazynist.dart
/*
 *
 * Refazynist is a Refreshable, Lazy and Animated List
 * Writen by Murat TAMCI aka THEMT.CO | http://themt.co
 * 2021/06/04 / v0.0.1
 *
 */

import 'dart:convert';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:shared_preferences/shared_preferences.dart';

typedef RefazynistErrorBuilder = Widget Function (BuildContext bContext, AsyncSnapshot bSnapshot);
typedef RefazynistEmptyBuilder = Widget Function (BuildContext bContext);
typedef RefazynistLoaderBuilder = Widget Function (BuildContext bContext, Animation<double> bAnimation);
typedef RefazynistOnLazy = Future<List<dynamic>> Function ();
typedef RefazynistItemBuilder = Widget Function (dynamic item, BuildContext bContext, int index, Animation<double> bAnimation, RefazynistCallType type);
typedef RefazynistRemovedItemBuilder = Widget Function (dynamic item, BuildContext bContext, int index, Animation<double> bAnimation, RefazynistCallType type);

/// Used for configure how [itemBuilder] or [removeItemBuilder] can be triggered
enum RefazynistCallType {
  /// The builder triggered can be single added
  item,

  /// The builder triggered can be mass added
  all,
}

/// Refazynist is a Refreshable, Lazy and Animated List
///
class Refazynist extends StatefulWidget {

  /// Creates a Refazynist
  ///
  /// The [onInit], [itemBuilder], [removedItemBuilder],
  /// [onLazy], [onRefresh], [sharedPreferencesName] arguments must be
  /// non-null.
  const Refazynist({
    required this.onInit,
    required this.itemBuilder,
    required this.removedItemBuilder,
    required this.onLazy,
    required this.onRefresh,
    required this.sharedPreferencesName,
    this.emptyBuilder = refazynistDefaultErrorBuilder,
    this.loaderBuilder = refazynistDefaultLoaderBuilder,
    this.insertDuration = const Duration(milliseconds: 250),
    this.removeDuration = const Duration(milliseconds: 250),
    this.sequentialInsert = false,
    this.sequentialRemove = false,
    Key? key
  }) : super(key: key);

  /// A name that's name for cache. Used for [SharedPreferences]
  final String sharedPreferencesName;

  /// A function that's called builder when list is empty
  final RefazynistEmptyBuilder emptyBuilder;

  /// itemBuilder for rendering on Animated List
  final RefazynistItemBuilder itemBuilder;

  /// itemBuilder for rendering on Animated List when item removed
  final RefazynistRemovedItemBuilder removedItemBuilder;

  /// A function that's called when lazy load is required
  final RefazynistOnLazy onLazy;

  /// A function that's called when swap to refresh
  final RefazynistOnLazy onRefresh;

  /// A function that's called when first run
  final RefazynistOnLazy onInit;

  /// A function that's called when loader rendering when lazy load required
  final RefazynistLoaderBuilder loaderBuilder;

  /// Used for how to be item inserting, sequential or same-time
  final bool sequentialInsert;

  /// Used for how to be item removing, sequential or same-time
  final bool sequentialRemove;

  /// Duration for animation when item inserted to list
  final Duration insertDuration;

  /// Duration for animation when item removed from list
  final Duration removeDuration;

  @override
  RefazynistState createState() => RefazynistState();
}

class _RefazynistItemParam {
  RefazynistCallType callType = RefazynistCallType.all;

  _RefazynistItemParam ({this.callType = RefazynistCallType.all});
}

class RefazynistState extends State<Refazynist> {

  GlobalKey<AnimatedListState> _animatedListKey = GlobalKey();
  List<dynamic> _items = <dynamic>[];
  List<_RefazynistItemParam> _params = <_RefazynistItemParam>[];

  /// A string that's store cursor in SharedPreferences
  String cursor = '';

  bool _end = false;
  Widget? _frontWidget;
  late Widget _animatedListWidget;
  
  void setTheState () {
    setState(() {

    });
  }

  void editItem (int index, dynamic item) {
    if (index >= 0 && index < _items.length) {
      _items[index] = item;
      _animatedListKey.currentState!.setState(() {

      });
      setState(() {

      });
    } else {
      print ('refazynist | editItem | index not in range');
    }
  }

  /// Get length of list
  int length () {
    return _items.length;
  }

  /// Remove an item by [index]
  /// [duration] used can be animation
  /// [disappear] user can be just remove it quickly, without animation
  Future<void> removeItem (int index, {Duration? duration, bool disappear = false}) async {

    if (index >= 0 && index < _items.length) {
      int oldLen = _items.length;
      await _removeItem(index, RefazynistCallType.item, duration: duration, disappear: disappear);
      oldLen--;
      if (_frontWidget == null && oldLen == 0) {
        _frontWidget = widget.emptyBuilder(context);
        setState(() {

        });
      }
      _setSharedPreferences();
    }
  }

  /// Insert an item by [item] at [index]
  /// callType used can be a type for itemBuilder or removeBuilder
  Future<void> insertItem (int index, dynamic item, {RefazynistCallType callType = RefazynistCallType.item}) async {
    _insertItem(index, item, callType: callType);
    _frontWidget = null;
    setState(() {

    });
    _setSharedPreferences();
  }

  /// Insert all items in [itemList] at [index]
  /// callType used can be a type for itemBuilder or removeBuilder
  Future<void> insertAllItem (int index, List<dynamic> itemList, {RefazynistCallType callType = RefazynistCallType.item}) async {
    _insertAllItem(index, itemList, callType: callType);
    _frontWidget = null;
    setState(() {

    });
    _setSharedPreferences();
  }

  Future<void> _setSharedPreferences () async {
    if (widget.sharedPreferencesName == '') return;

    SharedPreferences _prefs = await SharedPreferences.getInstance();

    await _prefs.setString(widget.sharedPreferencesName, jsonEncode({'cursor': cursor, '_end': !_loaderShowing, 'list': _items}));
  }

  Future<bool> _getSharedPreferences () async {

    if (widget.sharedPreferencesName == '') return false;

    SharedPreferences _prefs = await SharedPreferences.getInstance();

    String? spString = _prefs.getString(widget.sharedPreferencesName);

    if (spString == '') return false;

    if (spString != null) {
      Map<dynamic, dynamic> sp = jsonDecode(spString);

      if (sp.containsKey('cursor')) {
        cursor = sp['cursor'];
      }

      if (sp.containsKey('_end')) {
        _end = sp['_end'];
        if (_end) removeLoader();
        else addLoader();
      }

      if (sp.containsKey('list')) {
        List<dynamic> spList = sp['list'];

        if (spList.length > 0) {
          _insertAllItem(0, spList);
        }

        return true;
      }
    }

    return false;
  }

  Future<void> _refresh () async {

    if (!_play) return Future.value();

    cursor = "";

    List<dynamic> refreshList = await widget.onRefresh();

    await clear(sequentialRemove: widget.sequentialRemove);

    if (refreshList.length > 0) {

      _insertAllItem(_items.length, refreshList, sequentialInsert: widget.sequentialInsert, callType: RefazynistCallType.all);

      addLoader();
    }

    if (_items.length > 0) {
      _frontWidget = null;
    } else {
      _frontWidget = widget.emptyBuilder(context);
    }

    setState(() {

    });

    return Future.value(null);
  }
  
  @override
  void initState() {
    super.initState();

    _play = true;
    _loaderShowing = false;

    _animatedListWidget = AnimatedList(
      initialItemCount: _items.length + (_loaderShowing ? 1 : 0),
      key: _animatedListKey,
      itemBuilder: _itemBuilder,
    );

    _frontWidget = Center(
      child: CircularProgressIndicator(),
    );

    _onInit();
  }

  /// Clear the all item on list
  ///
  /// [sequentialRemove] used for remove process, true for ordered-time, false for same-time
  Future<void> clear ({bool sequentialRemove = false}) async {
    _clearRunning = true;

    //cursor = "";

    removeLoader();

    int oldLen = _items.length;

    if (sequentialRemove) {
      for (int i=0; i<oldLen; i++) {
        await _removeItem(0, RefazynistCallType.all);
      }
    } else {
      for (int i=0; i<oldLen; i++) {
        _removeItem(0, RefazynistCallType.all);
      }
    }


    _setSharedPreferences();

    await Future.delayed(widget.removeDuration);

    if (_items.length > 0) {
      _frontWidget = null;
    } else {
      _frontWidget = widget.emptyBuilder(context);
    }

    setState(() {

    });

    _clearRunning = false;
  }

  Future<void> _insertItem (index, dynamic item, {RefazynistCallType callType = RefazynistCallType.item}) async {
    assert (index >= 0);

    _items.insert(index, item);
    _params.insert(index, _RefazynistItemParam());
    _params[index].callType = callType;

    _animatedListKey.currentState!.insertItem(index, duration: widget.insertDuration);

    await Future.delayed(widget.insertDuration);

    return;
  }

  Future<void> _insertAllItem (int index, List<dynamic> list_items, {bool sequentialInsert = false, RefazynistCallType callType = RefazynistCallType.item}) async {
    if (sequentialInsert) {
      for (int i=0; i<list_items.length; i++) {
        await _insertItem(index+i, list_items[i], callType: callType);
      }
    } else {
      for (int i=0; i<list_items.length; i++) {
        _insertItem(index+i, list_items[i], callType: callType);
      }
    }
  }

  Future<dynamic> _removeItem (int index, RefazynistCallType callType, {Duration? duration, bool disappear = false}) async {
    assert (index >= 0);
    assert (index < _items.length);

    _params.removeAt(index);
    dynamic item = _items.removeAt(index);
    _animatedListKey.currentState!.removeItem(index, (context, animation) {
      if (disappear)
        return Container();
      else
        return widget.removedItemBuilder(item, context, index, animation, callType);
    }, duration: disappear ? Duration(seconds: 0) : duration??widget.removeDuration);

    if (disappear == false) await Future.delayed(duration??widget.removeDuration);



    return item;
  }

  /// Add loader on the end on list for indicate lazy load is working
  Future<void> addLoader () async {
    _loaderShowing = true;
    _onLazyRunning = false;
    if (_animatedListKey.currentState != null) {
      _animatedListKey.currentState!.insertItem(_items.length, duration: widget.insertDuration);
    }
  }

  /// Remove loader on the end on list when lazy load in progress
  Future<void> removeLoader () async {
    if (_loaderShowing) {
      _animatedListKey.currentState!.removeItem(_items.length, (riContext, rAnimation) {
        return widget.loaderBuilder(riContext, rAnimation);
      }, duration: widget.removeDuration);

      await Future.delayed(widget.removeDuration);

      _loaderShowing = false;
    }

  }

  bool _loaderShowing = false;
  bool _onLazyRunning = false;
  bool _clearRunning = false;
  bool _play = false;

  Widget _itemBuilder (BuildContext ibContext, int index, Animation<double> ibAnimation) {

    if (_loaderShowing && index >= _items.length) {

      // lazynin çağrılması için o an çalışmıyor olması gerek
      // ek olarak temizleme yapılmıyor olmalı. temizlerken eklemek saçma olur

      if (_play && _onLazyRunning == false && _clearRunning == false) {
        _onLazyRunning = true;

        Future<List<dynamic>> futureLazyList = widget.onLazy ();

        futureLazyList.then ((List<dynamic> lazyList) {

          if (!_play) return;
          if (_clearRunning) return; // temizlenme istendiyse ekleme yapmak anlamsız.

          if (lazyList.length == 0) {
            removeLoader().then((value) {
              _setSharedPreferences();
            });
          } else {
            _insertAllItem(_items.length, lazyList, sequentialInsert: widget.sequentialInsert, callType: RefazynistCallType.all);
            _setSharedPreferences();
          }


          _onLazyRunning = false;
        });
      }

      return widget.loaderBuilder(ibContext, ibAnimation);
    }

    if (index < _items.length) {
      return widget.itemBuilder (_items[index], ibContext, index, ibAnimation, _params[index].callType);
    }


    return Center(
      child: Text ("Overflow #$index"),
    );
  }

  Future<bool> _onInit () async {
    //print ('_onInit');

    if (!_play) return false;

    _loaderShowing = false;

    bool spResult = await _getSharedPreferences(); // kayıtlı veri varsa çek
    //bool spResult = false;

    // kayıtlı veri yoksa netten al
    if (!spResult) {
      List<dynamic> list = await widget.onInit();

      if (list.length > 0) {
        //_items.addAll(list);
        _insertAllItem(0, list, sequentialInsert: widget.sequentialInsert, callType: RefazynistCallType.all);

        await _setSharedPreferences();

      }
    }


    if (_items.length > 0) {
      _frontWidget = null;
    } else {
      _frontWidget = widget.emptyBuilder(context);
    }

    setState(() {

    });

    return true;
  }

  @override
  Widget build(BuildContext context) {

    if (_frontWidget != null) {
      return RefreshIndicator(
          child: Stack(
            children: [
              Opacity(
                opacity: 0,
                child: _animatedListWidget,
              ),
              _frontWidget!,
            ],
          ),
          onRefresh: _refresh
      );
    } else {
      return RefreshIndicator(
          child: _animatedListWidget,
          onRefresh: _refresh
      );
    }

  }
}

Widget refazynistDefaultErrorBuilder (BuildContext bContext) {
  return Stack(
    children: <Widget>[
      ListView(),
      Center(
        child: Wrap(
          children: [
            Column(
              children: [
                Icon(
                  Icons.warning_amber_rounded,
                  size: 60,
                  color: Colors.black26,
                ),
                Text ('Empty'),
              ],
            )
          ],
        ),
      ),
    ],
  );
}

Widget refazynistDefaultLoaderBuilder (BuildContext bContext, Animation<double> bAnimation) {
  return Center(
    child: FadeTransition(
        opacity: bAnimation,
        child: Padding(
          padding: EdgeInsets.all(20),
          child: CircularProgressIndicator(),
        )),
  );
}


example/lib/refazynist_demo_dismissible.dart
/*
 *
 * Dismissible Example | Refazynist is a Refreshable, Lazy and Animated List
 * Writen by Murat TAMCI aka THEMT.CO | http://themt.co
 * 2021/06/04
 *
 */

import 'package:flutter/material.dart';
import 'package:refazynist/refazynist.dart';
import 'package:shared_preferences/shared_preferences.dart';


void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Refazynist Dismissible Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: RefazynistDismissibleDemo(title: 'Refazynist Dismissible Demo'),
    );
  }
}

class RefazynistDismissibleDemo extends StatefulWidget {
  RefazynistDismissibleDemo({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  _RefazynistDismissibleDemoState createState() => _RefazynistDismissibleDemoState();
}

class _RefazynistDismissibleDemoState extends State<RefazynistDismissibleDemo> {

  GlobalKey<RefazynistState> refazynistKey = GlobalKey();

  int lazyCount = 5; // It's for lazy loading limit
  String sharedPreferencesName = 'dismissible_demo'; // It's for cache. Storing in Shared Preferences

  @override
  Widget build(BuildContext bContext) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Refazynist(
          key: refazynistKey,
          sharedPreferencesName: sharedPreferencesName,

          onInit: () async {
            return <dynamic>['Init item 1', 'Init item 2'];
          },

          emptyBuilder: (ewContext) {
            return Stack(
              children: <Widget>[
                ListView(),
                Center(
                  child: Wrap(
                    children: [
                      Column(
                        children: [
                          Icon(
                            Icons.warning_amber_rounded,
                            size: 60,
                            color: Colors.black26,
                          ),
                          Text ('Empty'),

                          Padding(padding: EdgeInsets.only(top: 20)),

                          ElevatedButton(
                            child: Text ('Create New'),

                            onPressed: () {
                              refazynistKey.currentState!.insertItem(0, 'Created item');
                            },
                          )
                        ],
                      )
                    ],
                  ),
                ),
              ],
            );
          },

          //
          // Refazynist: It's for refresh

          onRefresh: () async {
            lazyCount = 5; // reset lazy loading...

            await Future.delayed(Duration(seconds: 2)); // Fake internet delay

            return <dynamic>['Refresh item 0', 'Refresh item 1', 'Refresh item 2'];
          },

          //
          // Refazynist: It's for lazy load

          onLazy: () async {
            List<dynamic> lazyList = <dynamic>[];

            if (lazyCount > 0) {
              lazyCount--;
              lazyList.add ('Lazy item ' + (refazynistKey.currentState!.length() + 0).toString());
              lazyList.add ('Lazy item ' + (refazynistKey.currentState!.length() + 1).toString());
            }

            await Future.delayed(Duration(seconds: 1)); // Fake internet delay

            return lazyList;
          },

          //
          // Refazynist: itemBuilder

          itemBuilder: (item, ibContext, index, animation, type) {
            return Dismissible(
                key: Key(item),
                background: Container(
                  color: Theme.of(context).primaryColor,
                  child: Icon(Icons.delete_outline, color: Colors.white,),
                ),
                onDismissed: (_) {
                  refazynistKey.currentState!.removeItem(index, disappear: true);
                  setState(() {

                  });
                },
                child: FadeTransition(
                  opacity: animation,
                  child: Padding(
                    padding: EdgeInsets.only(left: 20, right: 20, top: 30, bottom: 30),
                    child: Row(
                      children: [
                        Expanded(child: Text ('$item')),
                        ElevatedButton(
                            onPressed: () {
                              refazynistKey.currentState!.removeItem(index);
                            },
                            child: Icon(Icons.delete_outline)
                        )
                      ],
                    ),
                  ),
                )
            );

          },

          //
          // Refazynist: removed ItemBuilder (need for Flutter's Animated List)

          removedItemBuilder: (item, ibContext, index, animation, type) {
            return FadeTransition(
              opacity: animation,
              child: Padding(
                padding: EdgeInsets.only(left: 20, right: 20, top: 30, bottom: 30),
                child: Row(
                  children: [
                    Expanded(child: Text ('$item')),
                    ElevatedButton(
                        onPressed: () {
                          refazynistKey.currentState!.removeItem(index);
                        },
                        child: Icon(Icons.delete_outline)
                    )
                  ],
                ),
              ),
            );
          }
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          showModalBottomSheet(
            context: context,
            builder: (sbsContext) {
              return ListView(
                children: [
                  ListTile(
                    title: Text('Add item to top'),
                    onTap: () {
                      refazynistKey.currentState!.insertItem(0, 'Added item ${refazynistKey.currentState!.length()}');
                    },
                  ),

                  ListTile(
                    title: Text('Remove item from top'),
                    onTap: () {
                      refazynistKey.currentState!.removeItem(0);
                    },
                  ),

                  ListTile(
                    title: Text('Clear The List'),
                    onTap: () {
                      refazynistKey.currentState!.clear();
                    },
                  ),

                  ListTile(
                    title: Text('Show Shared Preferences'),
                    onTap: () async {
                      SharedPreferences _prefs = await SharedPreferences.getInstance();

                      String? spString = _prefs.getString(sharedPreferencesName);

                      print ('Shared Preferences: $spString');

                      showModalBottomSheet(
                        context: sbsContext,
                        builder: (sbsContext2) {
                          return TextField(
                            maxLines: 15,
                            decoration: InputDecoration.collapsed(hintText: spString),
                          );
                        },
                      );
                    },
                  ),

                  ListTile(
                    title: Text('Clear Shared Preferences'),
                    onTap: () async {
                      SharedPreferences _prefs = await SharedPreferences.getInstance();

                      _prefs.setString(sharedPreferencesName, '');
                    },
                  )
                ],
              );
            },
          );
        },
        child: Icon(Icons.add),
      ),
    );
  }
}
